adds "How to create component" page (#875)
bbc/simorgh: docs/Creating-Components.md

data/schema.yaml describes the Article API definition for web and apps. We need to make it aware of our new component.

Your component is more than likely a new 'block', so you'll need to add it to the array of what blocks are allowed to look like:

```yaml
    blocks:
      type: object
      items:
        oneOf:
          - $ref: '#/components/schemas/altText'
          ... etc ...
          - $ref: '#/components/schemas/[your component name]'
      minItems: 2
```

You'll also need to define the block subtype itself:

```yaml
    blockquote:
      type: object
      required:
        - model
        - type
      properties:
        model:
          properties:
            blocks:
              $ref: '#/components/schemas/blocks'
          type: object
        type:
          enum:
            - blockquote
          type: string
```

The schema check currently only happens on local data. Eventually, we hope to run the schema validator at runtime.

### Update some local fixture data
Pick a JSON file under `data/test/news/articles/[id].json`, and:

1) add an example of your block somewhere in the `content.model.blocks` array.
2) add your new component to the `blockTypes` array.

> Run `npm run dev` and you should see your component at your article of choice, eg http://localhost:7080/news/articles/c0000000027o

Reviewer: Updating the link to be a valid one would be great http://localhost:7080/news/articles/c0000000001o

### Add a container
We've added our _component_, which should be kept as simple as possible. Now we need to create our _container_, which contains the business logic for mapping Optimo block data to the React parameters our component needs.

Add a new folder under `src/app/containers/[Component Name]/`. You will need:

* index.jsx - describes the mapping of Optimo block data to React parameters
* index.stories.jsx - describes the behaviour of your component under different data conditions

This step is quite complicated, so copy and paste from a similar example and tweak the code to your requirements.

As part of this work, you'll need to...

### Define your component propTypes
Under `src/app/models/propTypes/[Component Name (lowercase)]`, create an `index.jsx` which describes the propTypes your component needs. Import this into your `src/app/containers/[Component Name]/index.jsx`.

> ### Update the Article component

Reviewer: It would be great to clarify that integrating into the Article component should be done separately to the creation of the component and container. This is because we should conduct a full accessibility review / accessibility swarm as part of the integration ticket.

Author: Ok! I didn't know it should 👍 will work on it now. 

So for my understanding - 3 PRs:

1) presentational component
2) container component (and corresponding fixture data and schema updates)
 - does the accessibility swarm happen at this point? Does it then get merged?
3) integrate into Article

Is that right? Or does more happen in PR number 3?

Reviewer: Sorry, I should have explained more clearly. 
For PRs 1 & 2, we do some accessibility testing when the PR gets to the UX/Accessibility review column. 
We use this as a reference to guide the accessibility review step https://bbc-news.github.io/accessibility-news-and-you/accessibility-and-supported-assistive-technology

For PR 3, which is integrating into the Article, when the PR gets to the UX/Accessibility review column, we do a full accessibility swarm, with people from multiple disciplines - UX designers, devs, testers, BAs, etc. 

For this, we have a spreadsheet template that is currently in Google Drive. I will get a blank version to reference here in our documentation.

import { storiesOf } from '@storybook/react'; // eslint-disable-line import/no-extraneous-dependencies
import Blockquote from './index';

storiesOf('Blockquote', module).add('default - quote only', () => <Blockquote>To be or not to be, that is the question</Blockquote>);
storiesOf('Blockquote', module).add('quote with cite', () => (
    <Blockquote cite="William Shakespeare">To be or not to be, that is the question</Blockquote>
));
```

## Part 2: Create the container component
Once your component front-end is merged, you'll need to integrate it into the renderer. Let's work from the lowest level to the highest level:

### Add a test
In your component subdirectory, add a `index.test.jsx` to take a "snapshot" so that we can protect against regressions to the component. If the structure of the component changes, the unit test suite will fail.

```js
import React from 'react';
import { shouldMatchSnapshot } from '../../helpers/tests/testHelpers';
import Blockquote from './index';

describe('Blockquote', () => {
  shouldMatchSnapshot(
    'should render correctly',
    <Blockquote>To be or not to be, that is the question</Blockquote>,
  );
});
```

### Update the schema

> data/schema.yaml describes the Article API definition for web and apps. We need to make it aware of our new component.

Reviewer 2: ```suggestion
data/schema.yaml describes the Article API definition for web. We need to make it aware of our new component.
```

Currently apps are not using this schema

> Your component is more than likely a new 'block', so you'll need to add it to the array of what blocks are allowed to look like:

Reviewer 2: ```suggestion
Your component is more than likely a new 'block' in the data feed, so you'll need to add it to the array of which blocks the application should validate:
```

    blockquote:
      type: object
      required:
        - model
        - type
      properties:
        model:
          properties:
            blocks:
              $ref: '#/components/schemas/blocks'
          type: object
        type:
          enum:
            - blockquote
          type: string
```

The schema check currently only happens on local data. Eventually, we hope to run the schema validator at runtime.

### Update some local fixture data
Pick a JSON file under `data/test/news/articles/[id].json`, and:

1) add an example of your block somewhere in the `content.model.blocks` array.
2) add your new component to the `blockTypes` array.

Run `npm run dev` and you should see your component at your article of choice, eg http://localhost:7080/news/articles/c0000000001o

### Add a container
We've added our _component_, which should be kept as simple as possible. Now we need to create our _container_, which contains the business logic for mapping Optimo block data to the React parameters our component needs.

Add a new folder under `src/app/containers/[Component Name]/`. You will need:

* index.jsx - describes the mapping of Optimo block data to React parameters

> * index.stories.jsx - describes the behaviour of your component under different data conditions

Reviewer 2: ```suggestion

```

Containers should not have stories

Reviewer 2: ```suggestion
* index.test.jsx - creates "snapshots" of the component with the various different rendered outputs for the business logic in the container
```
Unless this was meant to be `index.test.jsx`

NB: we try to build "AMP-first", but this is difficult in practice because we are unable to inject the AMP dependency into Storybook. For now, check the AMP output manually in Part 2 of the process. Simorgh currently only supports one 'view' of components; components which require a different implementation for AMP must wait for the outcome of our [AMP/canonical component investigation](https://github.com/BBC-News/simorgh/issues/884).

Once done, make a PR - this first step will need code review and UX QA.

Example implementations below:

### index.jsx
```js
import styled from 'styled-components';

const Blockquote = styled.blockquote`
  background: #f9f9f9;
`;

export default Blockquote;
```

### index.stories.jsx
```js
import React from 'react';
import { storiesOf } from '@storybook/react'; // eslint-disable-line import/no-extraneous-dependencies
import Blockquote from './index';

storiesOf('Blockquote', module).add('default - quote only', () => <Blockquote>To be or not to be, that is the question</Blockquote>);
storiesOf('Blockquote', module).add('quote with cite', () => (
  <Blockquote cite="William Shakespeare">To be or not to be, that is the question</Blockquote>
));
```

> ## Part 2: Create the container component

Reviewer 2: Feels like this and line 108 (`### Add a container`) are the same thing. Maybe this heading and summary is in the wrong place?

Under `src/app/models/propTypes/[Component Name (lowercase)]`, create an `index.jsx` which describes the propTypes your component needs. Import this into your `src/app/containers/[Component Name]/index.jsx`.

## Part 3: Integrate the component into the Article
Before your component can be used in production, it must go through a full accessibility and UX review. These steps don't happen as much in the first two PRs as we don't want them to become a bottleneck for development.

### Build dependency into article renderer
As part of the final review process, you'll need to import your component's container and add it to the `componentsToRenderMain` object in `src/app/containers/Article/index.jsx`.

This feels like duplication of the schema - and it kind of is - but is currently our solution for preventing bundling _all components_ into the application JS.

### Testing
Before merge, if your new component requires integration testing it should be covered in the end-to-end (E2E) tests.

The test team is responsible for writing integration tests - [which live in Simorgh](https://github.com/BBC-News/simorgh/tree/latest/cypress) - covering your component's functionality.

### Publishing

> By now, you should have all the automatic PR checks passing, 2 manual code reviews Approved, and all the UX/accessibility/test QA completed.

Reviewer 3: ```suggestion
By now, you should have all the automatic PR checks passing, 2 manual code reviews approved, and all the UX/accessibility/test QA completed.
```

NB, the automatic checks are:

* CodeClimate (diff coverage, total coverage, code quality)
* Travis (runs the tests, [updates the storybook](https://simorghstorybook.now.sh/))
* Jenkins CI* (`bbc-news-simorgh` job - runs the tests).

> \* Currently we run tests on Jenkins with a view to moving away from Travis, so tests are currently run twice. We hope to simplify things soon.

Reviewer 3: Soz one small thing again 🙈 Talking to @sareh she wasn't aware moving away from travis is a thing, is it?

Reviewer 3: Discussion on slack with @sareh agreed we should leave this and change when we refactor the jenkins pipeline
